src/ipc/mem/PageStack.cc

```diff
@@ -10,79 +10,428 @@
+        toSubtractNext = toSubtract;
+        assert(value.right >= toSubtract);
+        value.left = value.left;
+        value.right -= toSubtract;
+    }
+    *valuePtr = value.pack();
+    return toSubtractNext;
+}
 
+/// accounts for an ID added to subtree in the given dir from the given position
 void
-Ipc::Mem::PageStackStorageSlot::take()
+Ipc::Mem::IdSet::innerPush(const Position pos, const NavigationDirection dir)
+{
+    // either left or right component will be true/1; the other will be false/0
+    const auto increment = IdSetInnerNode(dir == dirLeft, dir == dirRight).pack();
+    const auto previousValue = nodeAt(pos).fetch_add(increment);
+    // no overflows
+    assert(previousValue <= std::numeric_limits<Node>::max() - increment);
+}
+
+/// accounts for future ID removal from a subtree of the given position
+/// \returns the direction of the subtree chosen to relinquish the ID
+Ipc::Mem::IdSet::NavigationDirection
+Ipc::Mem::IdSet::innerPop(const Position pos)
 {
-    const auto nxt = nextOrMarker.exchange(TakenPage);
-    assert(nxt != TakenPage);
+    NavigationDirection direction = dirNone;
+
+    auto &node = nodeAt(pos);
+    auto oldValue = node.load();
+    IdSetInnerNode newValue;
+    do {
+        newValue = IdSetInnerNode::Unpack(oldValue);
+        if (newValue.left) {
+            --newValue.left;
+            direction = dirLeft;
+        } else if (newValue.right) {
+            --newValue.right;
+            direction = dirRight;
+        } else {
+            return dirEnd;
+        }
+    } while (!node.compare_exchange_weak(oldValue, newValue.pack()));
+
+    assert(direction == dirLeft || direction == dirRight);
+    return direction;
 }
 
+/// adds the given ID to the leaf node at the given position
 void
-Ipc::Mem::PageStackStorageSlot::put(const PointerOrMarker expected, const Pointer nxt)
+Ipc::Mem::IdSet::leafPush(const Position pos, const size_type id)
+{
+    const auto mask = Node(1) << (id % BitsPerLeaf);
+    const auto oldValue = nodeAt(pos).fetch_or(mask);
+    // this was a new entry
+    assert((oldValue & mask) == 0);
+}
+
+// TODO: After switching to C++20, use countr_zero() which may compile to a
+// single TZCNT assembly instruction on modern CPUs.
+/// a temporary C++20 countr_zero() replacement
+static inline
+int trailingZeros(uint64_t x)
 {
-    assert(nxt != TakenPage);
-    const auto old = nextOrMarker.exchange(nxt);
-    assert(old == expected);
+    if (!x)
+        return 64;
+    int count = 0;
+    for (uint64_t mask = 1; !(x & mask); mask <<= 1)
+        ++count;
+    return count;
+}
+
+/// extracts and returns an ID from the leaf node at the given position
+Ipc::Mem::IdSet::size_type
+Ipc::Mem::IdSet::leafPop(const Position pos)
+{
+    auto &node = nodeAt(pos);
+    auto oldValue = node.load();
+    Node newValue;
+    do {
+        assert(oldValue > 0);
+        const auto mask = oldValue - 1; // flips (the last 1 and trailing 0s)
+        newValue = oldValue & mask;
```

Comment:
I would also add an extra comment, making it easier to understand the intention:
```suggestion
        newValue = oldValue & mask; // flips the rightmost 1
```

Reply:
Good idea. Done in 646dee4.